### DISCUSSION TITLE
Feature Request for Max Session Time Limits and Inactive Turn Off Window

### DISCUSSION THREAD

**Discussion by hockeyiscool19** [fastapi/fastapi#10664] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# Minimal Example
from fastapi import FastAPI, Request
from starlette.middleware.base import BaseHTTPMiddleware
from datetime import datetime, timedelta
import asyncio
import signal

from typing import Union
from fastapi import FastAPI
from dotenv import load_dotenv
from app.utils.gmail import GMAIL
from app.utils.firebase import FIRE
import os

load_dotenv()
app = FastAPI()
last_request_time = datetime.now()

@app.middleware("http")
async def update_last_request_time(request: Request, call_next):
    global last_request_time
    last_request_time = datetime.now()
    response = await call_next(request)
    return response

def shutdown_server():
    os.kill(os.getpid(), signal.SIGINT)

async def check_inactivity():
    global last_request_time
    idle_time = timedelta(minutes=5)
    while True:
        print("Checking activity")
        await asyncio.sleep(10)
        if datetime.now() - last_request_time > idle_time:
            print("Shutting down due to inactivity.")
            shutdown_server()

loop = asyncio.get_event_loop()
loop.create_task(check_inactivity())

@app.get("/")
def read_root():
    return {"Hello": "Welcome to Jordan's API"}
```

### Feature Request
This is a feature request to simplify turning off FastAPI due to inactivity or no requests. What I am proposing is to create a way so that people can shut of FastAPI based on a few parameters. The options: 

1. Max Session: you can set the API to turn after a certain number of seconds. 
2. Inactivity: your API will shut off without clicking in SwaggerUI or requesting data. 
3. Both: you can set both options. That is, you can have a max sessions and also an inactivity window. 

### Function Calling
Ideally, you would call the function like:
```
from FastApi import session_limits
app = FastAPI(...)
session_limits = session_limits(max_session_time: int, inactivity_window: int)
```
Behind the scenes, this function would implement a middleware route, similar to this example:
```
from fastapi import FastAPI, Request
from starlette.middleware.base import BaseHTTPMiddleware
from datetime import datetime, timedelta
import asyncio
import signal

last_request_time = datetime.now()

@app.middleware("http")
async def update_last_request_time(request: Request, call_next):
    global last_request_time
    last_request_time = datetime.now()
    response = await call_next(request)
    return response

def shutdown_server():
    os.kill(os.getpid(), signal.SIGINT)

async def check_inactivity():
    global last_request_time
    idle_time = timedelta(minutes=5)
    while True:
        print("Checking activity")
        await asyncio.sleep(10)
        if datetime.now() - last_request_time > idle_time:
            print("Shutting down due to inactivity.")
            shutdown_server()

loop = asyncio.get_event_loop()
loop.create_task(check_inactivity())
```
Going through step by step, this example implements multi-threading. Every 10 seconds, the shut-down thread checks if the app had reached its max session time. The code continues if it does not reach the limit and stops if it does reach the limit. 

More abstractly, this feature would simplify these functions to be used in one line, with optional parameters. More details in this regard would appear in a potential PR.

### Why would this feature be useful and in the spirit of FastAPI?

One of the benefits of using FastAPI is that it is great for individual developers. These developers, however, should not have to worry about making sure their applications turn off when they need to. Not turning off an API can be exceedingly costly--on the order of hundreds of dollars every month while running on a cloud environment. (Trust me, I have made this mistake.) 

This feature would simplify a small yet expensive problem for many developers who leave their services on too long. It would provide a useful abstraction for developers looking to host personal projects, limitting their running costs. This is especially useful in the context of serverless computing, such as Cloud Run on GCP. For my personal websites, I chose not to spend five dollars a month on services such as Vercel. With this middleware shut-down solution, I spend pennies per month hosting my website. 

### Operating System

Linux

### Operating System Details

Using Linux locally and a Docker Container. https://github.com/hockeyiscool19/mygpt/blob/main/app/main.py

### FastAPI Version

0.104.1

### Pydantic Version

2.4.2

### Python Version

3.10.11

**Comment by jgould22** [comment, marked as answer]
Generally this is better handled by the cloud provider in question, not the framework.

It is a separation of concerns thing. FastAPI is a web development framework, not a cloud cost manager. 

Better to use aws auto scaling, autoscale in Azure, or whatever GCP is calling that feature.

**Reply by hockeyiscool19** [reply]
Interesting, I tried to look for a feature like that on GCP and could not find it! I'll take a second glance, but is there anywhere you recommend?

**Reply by jgould22** [reply]
Honestly I have very little experience with GCP but would be pretty surprised if they did not have a feature like this.
